### PR TITLE
test: migrate rust_version and rustc* to snapbox

### PR DESCRIPTION
**Pull request by eth3lbert**
### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/14039.

Migrate following to snapbox:

- `tests/testsuite/rust_version.rs`
- `tests/testsuite/rustc.rs`
- `tests/testsuite/rustc_info_cache.rs`

### CODE REVIEW

File: tests/testsuite/rustc.rs [hunk 1]
```diff
@@ -462,17 +427,17 @@ fn build_only_bar_dependency() {
...
-[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [..]
-",
-        )
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
         .run();
 }
 
+#[allow(deprecated)]
```
Comment (eth3lbert) [inline tests/testsuite/rustc.rs:440]: `with_stderr_does_not_contain` is being used in this function.

File: tests/testsuite/rustc.rs [hunk 2]
```diff
@@ -622,6 +589,7 @@ fn rustc_with_other_profile() {
     p.cargo("rustc --profile test").run();
 }
 
+#[allow(deprecated)]
```
Comment (eth3lbert) [inline tests/testsuite/rustc.rs:592]: `with_stderr_does_not_contain` is being used in this function.

File: tests/testsuite/rustc_info_cache.rs [hunk 3]
```diff
@@ -10,6 +8,7 @@ const MISS: &str = "[..] rustc info cache miss[..]";
 const HIT: &str = "[..]rustc info cache hit[..]";
 const UPDATE: &str = "[..]updated rustc info cache[..]";
 
+#[allow(deprecated)]
```
Comment (eth3lbert) [inline tests/testsuite/rustc_info_cache.rs:11]: `with_stderr_does_not_contain` is being used in this function.

File: tests/testsuite/rustc_info_cache.rs [hunk 4]
```diff
@@ -105,6 +104,7 @@ fn rustc_info_cache() {
         .run();
 }
 
+#[allow(deprecated)]
```
Comment (eth3lbert) [inline tests/testsuite/rustc_info_cache.rs:107]: `with_stderr_does_not_contain` is being used in this function.

File: tests/testsuite/rustc_info_cache.rs
Comment (eth3lbert) [inline tests/testsuite/rustc_info_cache.rs:1]: Based on the assertion that shows a pair of `contain` with `does_not_contain`, I'm inclined to keep it as is.

File: tests/testsuite/rustc.rs [hunk 5]
```diff
@@ -736,21 +708,36 @@ fn rustc_with_print_cfg_multiple_targets() {
...
-        .with_stdout_contains("target_pointer_width=\"64\"")
-        .with_stdout_contains("target_vendor=\"pc\"")
-        .with_stdout_contains("windows")
-        .with_stdout_contains("target_env=\"gnu\"")
-        .with_stdout_contains("target_family=\"unix\"")
-        .with_stdout_contains("target_pointer_width=\"32\"")
-        .with_stdout_contains("target_vendor=\"unknown\"")
-        .with_stdout_contains("target_os=\"linux\"")
-        .with_stdout_contains("unix")
+        .with_stdout_data(str![[r#"
```
Comment (weihanglo) [inline tests/testsuite/rustc.rs:711]: The order is not so importnat, should we use `.unordered()` and just one `...` instead?

File: tests/testsuite/rustc_info_cache.rs
Comment (weihanglo) [reply tests/testsuite/rustc_info_cache.rs:1]: How bad is it if we compare the full output? I think we may need to add a comment for each cargo invocation for the expected hit/miss/update output.

I do want to kill `does_not_contain` as it is even more unreliable.

File: tests/testsuite/rustc.rs (hunk 5 above)
Comment (eth3lbert) [reply tests/testsuite/rustc.rs:711]: Ok, this also works and make sense to me!

File: tests/testsuite/rustc_info_cache.rs
Comment (eth3lbert) [reply tests/testsuite/rustc_info_cache.rs:1]: Sounds good!  I'll give it a shot and see what results we can get.

Comment (eth3lbert) [reply tests/testsuite/rustc_info_cache.rs:1]: Oh, it seems like there are extra log messages related to `rustup`.

Comment (weihanglo) [reply tests/testsuite/rustc_info_cache.rs:1]: That is terribly bad. Let's keep the old tests.

Comment (eth3lbert) [reply tests/testsuite/rustc_info_cache.rs:1]: done reverting!

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Looks great. Thank you for the migration.

@bors r+

**Pull request merged**